plugin/pulpcore/plugin/download/asyncio/group.py

```diff
@@ -0,0 +1,190 @@
+
+    def schedule_from_iterator(self, group_iterator, parallel_group_limit=50):
+        """
+        Schedule groups to be downloaded using an iterator provided by the user.
+
+        When scheduling groups from the iterator, the `parallel_group_limit` argument defines the
+        number of groups that are handled in parallel. Once one group completes and is returned to
+        the user, another one is scheduled which maintains the number of parallel groups until the
+        iterator is exhausted.
+
+        When the group_iterator is a generator, this technique limits the effective memory used by
+        the DeferredDownload objects and their associated downloaders.
+
+        Additional calls to this will correctly schedule additional iterators, but only the first
+        call will specify the `parallel_group_limit`.
+
+        Args:
+            group_iterator (iterable): An iterable of tuples (id, [DeferredArtifact]). Each tuple is
+                one group with `id` being returned to the caller when all `DeferredArtifact.url`
+                downloads are complete.
+            parallel_group_limit (int): The number of groups to be handled in parallel at any time.
+        """
+        if self.group_iterator:
+            # This is not the first call to schedule an iterable
+            self.group_iterator = chain(self.group_iterator, group_iterator)
+            return
+
+        self.group_iterator = group_iterator
+        for i in range(parallel_group_limit):
+            with suppress(StopIteration):
+                self.schedule_group(**next(self.group_iterator))
+
+    def schedule_group(self, id, deferred_artifacts):
+        """
+        Schedules a group of `deferred_artifacts` for downloading, referred to by an `id`.
+
+        Args:
+            id (hashable): This id is handed back to the caller when all files in the group are
+                downloaded.
+            deferred_artifacts (list): A list of :class:`pulpcore.plugin.models.DeferredArtifact`
+                instances that have not been saved to the database.
+        """
+        group = Group(id, deferred_artifacts)
+        self.groups_not_done.append(group)
+        for url in group.urls:
+            if len(self.urls[url]) == 0:
+                # This is the first time we've seen this url so make a downloader
+                downloader_for_url = self.download_factory.build(url)
+                self.downloads_not_done.add(downloader_for_url)
+            self.urls[url].append(group)
+
+    def _find_and_remove_done_group(self):
+        for index, content_unit in enumerate(self.groups_not_done):
+            if content_unit.done:
+                self.groups_not_done.pop(index)
+                return content_unit
+
+    def __iter__(self):
+        return self
+
+    def __next__(self):
+        """
+        Returns:
+            tuple: The first element is the `id` first specified by the user. The second element is
```

Comment:
I *think* this actually returns `(id, {DeferredArtifact: DownloadResult})`.  Right?

Reply:
It's a little odd that the input is Group(s) and output is something *group-like*.  What do you think about returning the finished `Group` and letting the user inspect the group.

Something like:
```
for group in downloader:
    for deferred_artifact, download_result in group.finished_downloads:
        # do something.
```
Or,  may just go ahead and create the Artifact for them since the group downloader is artifact oriented.

Something like:
```
for group in downloader:
    for deferred_artifact, downloaded_artifact in group.downloaded_artifacts:
        # do something.
```

Reply:
I'm switching it to return `Group` and also to have it be `schedule_group(group)`.

Reply:
This is a good idea in making the switch a good amount of code was removed from the PR.